distributed/comm/utils.py

```diff
@@ -19,6 +21,63 @@
     OFFLOAD_THRESHOLD = parse_bytes(OFFLOAD_THRESHOLD)
 
 
+def is_msgpack_serializable(msg) -> bool:
```

Comment:
This approach feels brittle to me.  It's like we're reverse engineering our own protocol in order to determine if something is fast or not.  This feels off.

cc @jcrist

Reply:
Another option we discussed is maybe having a class `MsgPackSerializable`, which the user sending the message could place their message in. We could then just check for that class and use MsgPack on it in the fast path.

Reply:
This is to infer if a message contains objects that need to pass through the slow path of the serializers? I suspect most handlers are consistent in whether they send only simple objects or complext objects. A `bool` kwarg passed through the proper rpc/`write` methods would be much simpler and likely sufficient.

Pseudocode, not suggesting `comm.write` is the correct place for this, or that the kwarg name is a good one.

```python
await comm.write({"simple": "message"}, use_serializers=False)  # skip the serializers

await comm.write(some_complex_unknown_object, use_serializers=True)  # use the serializers
```

Reply:
Yeah I think Matt played with a flag to `write` and had inconsistent results. IOW it's not clear the message is MsgPack serializable by the time it reaches `write`. Please feel free to correct me Matt

This becomes more apparent when looking at the number of things using `BatchedSend` for example. I think we need to flag the message before we reach `write` if we don't want to handle it this way.

Reply:
I'm not sure that I ever tried this explicitly, but yeah I can imagine that it would get tricky, especially with BatchedSend because we might have a mix.  

We could include the `"fast": True` keyvalue pair in the message itself and pop it off as it goes by.  For lists we would have to iterate through the entire list.  It's not my favorite solution, but it seems more explicit than what is here.

Reply:
> Another option we discussed is maybe having a class MsgPackSerializable, which the user sending the message could place their message in.

The other way around might make more sense.  It looks like there are around 10-15 calls to `to_serialize` in the codebase.  Maybe we special case these calls instead of the pure MsgPack ones (which are more common and lighter weight).

Reply:
I agree, this is a hacky solution that is mainly for investigating the performance overhead of `extract_serialize`. @jakirkham @mrocklin can I get one of you to benchmark the changes with the Cythonized version?

If this PR makes a significant difference on a real workload, I think we should unify all serialization into a single function that pass through the data once.

Reply:
@quasiben did a run and posted info in issue ( https://github.com/quasiben/dask-scheduler-performance/issues/102 )

Reply:
More recent results in issue ( https://github.com/quasiben/dask-scheduler-performance/issues/103 ). Please ignore the previous one

Reply:
Overall time spent in `write` dropped from `5.25%` to `4.39%`. So pretty significant actually.

Would also note that all messages that are not MsgPack compatible are going through this extra step. So this improvement is seen in spite of adding overhead in some cases